platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/mappingToJvm/RuntimeFunctionResolver.kt

```diff
@@ -39,13 +44,44 @@ class MemberFunctionResolver(private val configureLambdaHandler: ConfigureLambda
         val parameterBindingStub = schemaFunction.parameters.associateWith { Any() }
         val hasConfigureLambda = (schemaFunction.semantics as? FunctionSemantics.ConfigureSemantics)?.configureBlockRequirement?.allows ?: false
 
-        receiverClass.memberFunctions.forEach { function ->
-            if (function.name == schemaFunction.simpleName && FunctionBinding.convertBinding(function, Any(), parameterBindingStub, hasConfigureLambda, configureLambdaHandler) != null) {
-                return RuntimeFunctionResolver.Resolution.Resolved(ReflectionFunction(function, configureLambdaHandler))
+        val resolutions = receiverClass.memberFunctions
+            .filter { function -> function.name == schemaFunction.simpleName && FunctionBinding.convertBinding(function, Any(), parameterBindingStub, hasConfigureLambda, configureLambdaHandler) != null }
+            .toList()
+
+        return when {
+            resolutions.isEmpty() -> RuntimeFunctionResolver.Resolution.Unresolved
+            resolutions.size == 1 -> RuntimeFunctionResolver.Resolution.Resolved(ReflectionFunction(resolutions[0], configureLambdaHandler))
+            else -> {
+                val refinedResolutions = resolutions.filter { function -> parametersMatch(function, schemaFunction) }
+                val finalResolution = if (refinedResolutions.size == 1) refinedResolutions[0] else
+                    error("Failed disambiguating between following functions: ${resolutions.joinToString(prefix = "\n\t", separator = "\n\t") { f -> f.toString() }}")
+                return RuntimeFunctionResolver.Resolution.Resolved(ReflectionFunction(finalResolution, configureLambdaHandler))
             }
         }
 
-        return RuntimeFunctionResolver.Resolution.Unresolved
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    private fun parametersMatch(function: KFunction<*>, schemaFunction: SchemaFunction): Boolean {
+
+        val actualParameters = function.parameters.subList(if (function.instanceParameter == null) 0 else 1, function.parameters.size)
+        return if (actualParameters.size == schemaFunction.parameters.size) {
+            return actualParameters.zip(schemaFunction.parameters).all { (kp, dp) ->
```

Comment:
`return` is unnecessary here, as it is the last expression in the `return if`'s branch.